core/lib/mini_merkle_tree/src/tests.rs

```diff
@@ -156,24 +158,79 @@ fn verify_merkle_proof(
     assert_eq!(hash, merkle_root);
 }
 
+fn verify_range_merkle_proof(
+    items: &[[u8; 88]],
+    mut start_index: usize,
+    start_path: &[Option<H256>],
+    end_path: &[Option<H256>],
+    merkle_root: H256,
+) {
+    assert_eq!(start_path.len(), end_path.len());
+
+    let hasher = KeccakHasher;
+    let mut hashes: VecDeque<_> = items.iter().map(|item| hasher.hash_bytes(item)).collect();
+
+    for (start_item, end_item) in start_path.iter().zip(end_path.iter()) {
+        if start_index % 2 == 1 {
+            hashes.push_front(start_item.unwrap());
+        } else {
+            assert_eq!(start_item, &None);
+        }
+        if hashes.len() % 2 == 1 {
+            hashes.push_back(end_item.unwrap());
+        } else {
+            assert_eq!(end_item, &None);
+        }
+
+        let next_level_len = hashes.len() / 2;
+        for i in 0..next_level_len {
+            hashes[i] = hasher.compress(&hashes[2 * i], &hashes[2 * i + 1]);
+        }
+
+        hashes.truncate(next_level_len);
+        start_index /= 2;
+    }
+
+    assert_eq!(hashes[0], merkle_root);
+}
+
 #[test]
 fn merkle_proofs_are_valid_in_small_tree() {
     let leaves = (1_u8..=50).map(|byte| [byte; 88]);
     let tree = MiniMerkleTree::new(leaves.clone(), None);
 
     for (i, item) in leaves.enumerate() {
-        let (merkle_root, path) = tree.clone().merkle_root_and_path(i);
+        let (merkle_root, path) = tree.merkle_root_and_path(i);
         verify_merkle_proof(&item, i, 50, &path, merkle_root);
     }
 }
 
+#[test]
+fn merkle_proofs_are_valid_for_ranges() {
+    let mut leaves: Vec<_> = (1_u8..=50).map(|byte| [byte; 88]).collect();
+    let mut tree = MiniMerkleTree::new(leaves.clone().into_iter(), None);
+
+    for i in 1..=50 {
+        let (merkle_root, start_path, end_path) = tree.merkle_root_and_paths_for_range(i);
+        verify_range_merkle_proof(&leaves[..i], 0, &start_path, &end_path, merkle_root);
+    }
+
+    tree.trim_start(25);
+    leaves.drain(..25);
+
+    for i in 1..=25 {
+        let (merkle_root, start_path, end_path) = tree.merkle_root_and_paths_for_range(i);
+        verify_range_merkle_proof(&leaves[..i], 25, &start_path, &end_path, merkle_root);
+    }
```

Comment:
I'd advise to test more thoroughly, e.g. like this

```rust
let mut start_index = 0;
for trimmed_count in 1..10 {
    tree.trim_start(trimmed_count);
    leaves.drain(..trimmed_count);
    start_index += trimmed_count;
    let tree_len = tree.hashes.len();

    for i in 1..=tree_len {
        let (merkle_root, start_path, end_path) = tree.merkle_root_and_paths_for_range(i);
        verify_range_merkle_proof(
            &leaves[..i],
            start_index,
            &start_path,
            &end_path,
            merkle_root,
        );
    }
}
```

I.e., remove 1, 2, 3, ..., 9 leaves at a time and test Merkle proofs for all remaining leaves in the tree.